[components] DbtProjectComponent

OwenKephart opened the pull request:
## Summary & Motivation

As title -- this adds a component that loads a dbt project into dbt assets

## How I Tested These Changes

```
# this creates a components/my_dbt_project directory, empty except for a defs.yml file
dg component generate dbt_project my_dbt_project

# get a sample project into this directory
cp -R path/to/jaffle_shop components/my_dbt_project
```

then, you edit `defs.yml` to be:

```yaml
component_type: ...

component_params:
  dbt:
    project_dir: my_dbt_project 
```

This is at least a bit complicated at the moment, so the next step here is to let all of that business get sorted in the generate step, by adding parameterization to it

## Changelog

NOCHANGELOG

OwenKephart commented:
* **#26281** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26281?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26281?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

OwenKephart commented on examples/experimental/dagster-components/dagster_components/impls/dbt_project.py:
```diff
@@ -0,0 +1,41 @@
+    def from_decl_node(cls, context: ComponentLoadContext, decl_node: ComponentDeclNode) -> Self:
+        assert isinstance(decl_node, YamlComponentDecl)
+
+        # all paths should be resolved relative to the directory we're in
```
more color: this lets us configure the dbt resource with `project_dir: jaffle_shop` and have that be relative to the file where that yaml is defined.

OwenKephart commented on examples/experimental/dagster-components/dagster_components/impls/dbt_project.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class DbtProjectComponent(Component):
+    params_schema = DbtCliResource
```
Another case where I found it useful to slap the resource in as the params schema -- this will probably get nested underneath another key eventually, i.e.:

```yaml
component_type: ...

component_params:
  dbt_resource:
     project_dir: jaffle_shop
  other_stuff: ...
```

schrockn replied:
Yes I think as convention it is critical to always have a top-level schema object that is extendible. The resource can be attached to that.

OwenKephart commented on examples/experimental/dagster-components/dagster_components/impls/dbt_project.py:
```diff
@@ -0,0 +1,41 @@
+
+    def build_defs(self, context: ComponentLoadContext) -> Definitions:
+        project = DbtProject(self.dbt_resource.project_dir)
+        project.prepare_if_dev()
```
couldn't decide if this belonged in the `from_decl_node` bit or here, but regardless abstracting this kinda nasty bit away seems like a big win

schrockn replied:
I think `build_defs` is the right call for now but we definitely need to formalize our component lifecycle so the answer is obvious.

schrockn reviewed (commented):
Can you make sure to add a run of the cmdline that generates the scaffolding and then also demonstrate what the files in the codebase are after the scaffold in the test plan?

schrockn reviewed (dismissed):
This is getting exciting.

* Resource shouldn't be top-level schema per comment.
* I think need two separate components, a jaffle_shop one and an empty project.
* I would also like to plop the integrations in extras or something, so we can demonstrate the process of incrementally installing modules and then getting more components in the component registry. This can be a follow up. cc: @smackesey for context

OwenKephart commented:
@schrockn 

> Resource shouldn't be top-level schema per comment.

Updated -- went with `dbt` as the key for two reasons:
1. to someone unfamiliar with the intricacies of the dagster resource system, calling this (required, and important) section `dbt_resource` would be pretty mysterious
2. it's a defensible name, as we use `dbt: DbtCliResource` in the decorated function in all of our examples

> I think need two separate components, a jaffle_shop one and an empty project.

Talked a bit about this offline, updated the test plan to demonstrate what's actually going on, but the gist is that we should do more in the generate_files() step than we currently do, but this needs to be controlled by some sort of parameterization scheme, so going to do that upstack as that's more of a core framework addition.

> I would also like to plop the integrations in extras or something

Yeah definitely -- I already had sling in extras, added a dbt extras as well. Just haven't added the code to inspect what's currently installed and use that to determine what gets added to the__component_registry__

schrockn reviewed (approved):
lfg

OwenKephart commented:
### Merge activity

* **Dec 5, 10:36 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/26281).
